src/main/java/picard/sam/markduplicates/UmiAwareDuplicateSetIterator.java

```diff
@@ -0,0 +1,251 @@
+        }
+        umiGraph.processNodes();
+
+        Map<Integer, List<UmiSequence>> umisByDuplicateSet = umiGraph.getUmisByDuplicateSet();
+
+        // Construct DuplicateSetList
+        List<DuplicateSet> duplicateSetList = new ArrayList<>();
+        for(int i = 0;i < umiGraph.duplicateSetCounter;i++) {
+            DuplicateSet e = new DuplicateSet();
+            duplicateSetList.add(e);
+        }
+
+        for(SAMRecord rec : records) {
+            String umi = records.get(records.indexOf(rec)).getStringAttribute(umiTag);
+
+            // Figure out which group this belongs to
+            int recordGroup = umiGraph.node.get(umi).duplicateSet;
+            duplicateSetList.get(recordGroup).add(records.get(records.indexOf(rec)));
+        }
+
+        // Optionally add the inferred Umi
+        if(addInferredUmi) {
+            for(DuplicateSet ds : duplicateSetList) {
+                // For each duplicate set identify the most common UMI
+                List<SAMRecord> recordsFromDuplicateSet = ds.getRecords();
+
+                // Assign the inferred UMI to all reads in the current group
+                for(SAMRecord record : recordsFromDuplicateSet) {
+                    record.setAttribute(inferredUmiTag, umiGraph.getInferredUmiFromOriginalSequence(record.getStringAttribute(umiTag)));
+                }
+            }
+        }
+
+        nextSetsIterator = duplicateSetList.iterator();
+    }
+
+    private int getEditDistance(final String s1, final String s2) {
+        if(s1 == null || s2 == null) {
+            throw new PicardException("Attempt to compare two incomparable UMIs.  At least one of the UMIs was null.");
+        }
+        if(s1.length() != s2.length()) {
+            throw new PicardException("Barcode " + s1 + " and " + s2 + " do not have matching lengths.");
+        }
+        int count = 0;
+        for(int i = 0;i < s1.length();i++) {
+            if(s1.charAt(i) != s2.charAt(i)) {
+                count++;
+            }
+        }
+        return count;
+    }
+
+    private class UmiSequence {
+        private final String sequence;
+        private final Long occupancy;
+        private String inferredSequence;
+        private Integer duplicateSet;
+
+        UmiSequence(final String sequence, final Long occupancy) {
+            this.sequence = sequence;
+            this.occupancy = occupancy;
+        }
+    }
+
+    private class UmiGraph {
+        private int duplicateSetCounter = 0;
+        private final Map<String, UmiSequence> node = new HashMap<>();
+
+        void addNode(final UmiSequence umi) {
+            node.put(umi.sequence, umi);
+        }
+
+        void processNodes() {
+            boolean isFirstEntry = true;
+            for(Map.Entry<String, UmiSequence> umi1 : node.entrySet()) {
+
+                // The first entry is a special case.  We automatically assign
+                // the first UMI we see to the first duplicate set.
+                if(isFirstEntry) {
```

Comment:
not sure why this is needed. the loop below will be empty, and then the if statement below that will kick in...no?